### PR TITLE
metrics: port network-latency test to CC3.x

### LINKED ISSUE #320: Port network-latency test to CC3.x
Part of #149.

### PR DESCRIPTION
**Pull request by bwarden**
Move network-latency.sh from wip folder, add it to run_all_metrics.sh.

Fixes #320

Signed-off-by: Brett T. Warden <brett.t.warden@intel.com>

### CODE REVIEW

**Comment by bwarden** [comment]
Ah, I suppose I should accomplish the results reporting in this PR as well.

**Comment by grahamwhaley** [comment]
Hi @bwarden Fundamentals look OK. I have a request though. I didn't have CC3.x installed on the machine, so I did a test run with `RUNTIME=runc`. I think all the tests should be runnable with any runtime (unless they have some runtime specific operations coded into them...). Doing that showed the network test are not laid out to run with any/other runtimes. I think we should fix that now before we proceed. Two things need fixing I think.

1) the `docker run` invocations don't currently take a runtime argument. They probably should. I did a quick test:

```
@@ -37,7 +37,7 @@ function start_server {
        local server_name="$1"
        local image="$2"
        local server_command="$3"
-       $DOCKER_EXE run -d --name=${server_name} ${image} sh -c "${server_command}" > /dev/null
+       $DOCKER_EXE run -d --runtime=${RUNTIME} --name=${server_name} ${image} sh -c "${server_command}" > /dev/null
        local server_address=$($DOCKER_EXE inspect --format "{{.NetworkSettings.IPAddress}}" ${server_name})
        echo "$server_address"
 }
@@ -49,7 +49,7 @@ function start_client {
        local client_name="$2"
        local image="$3"
        local client_command="$4"
-       $DOCKER_EXE run ${extra_args} --name=${client_name} ${image} sh -c "${client_command}"
+       $DOCKER_EXE run ${extra_args} --runtime=${RUNTIME} --name=${client_name} ${image} sh -c "${client_command}"
 }
```

2) The `runtime_docker` function checks that the required/expected runtime is the current docker default. I don't think that is correct. It does not have to be the default, particularly if we fix the missing `--runtime` arguments as per (1). What we could do is change that function to check that the requested runtime is actually installed in docker - that makes more sense, that it is actually available.

Make sense?

**Comment by bwarden** [comment]
Sure, I can rewrite the function to check for runtime in, e.g. "Runtimes: cc-runtime runc", and change the "$DOCKER_EXE run" calls to add the runtime.

**Comment by bwarden** [comment]
Function init_env() checks whether to restart cc-proxy by looking at docker's Default Runtime -- should this also be changed?

`    # Restart services
    systemctl restart docker
    runtime=$(docker info | grep -w "Default Runtime:" \
            | awk '{print $3}')

    if [ "$runtime" == "cor" ];then
        systemctl restart cc-proxy
    fi`

**Comment by grahamwhaley** [comment]
I suspect that should also be changed (proxy restart), yes.
So, we check that $RUNTIME is in the list of installed runtimes. We don't actually have to (or need to) hard code any of the known names (cor, cc-runtime, runc etc.) I think - we can make this pretty dynamic and flexible

File: metrics/lib/common.bash [hunk 1]
```diff
@@ -163,10 +163,16 @@ function get_qemu_path(){
...
+	runtimes=($(${DOCKER_EXE} info 2>/dev/null | grep "^Runtimes" \
+		| cut -d: -f2))
+	for runtime in ${runtimes[@]}; do
+		if [[ "${RUNTIME}" == "${runtime}" ]]; then
+			found=1
+			break;
+		fi
+	done
+	if [[ ! $found ]]; then
+		die "Selected runtime ${RUNTIME} not installed in docker (currently: ${runtimes[@]})"
```
Comment (grahamwhaley) [inline metrics/lib/common.bash]: nice. thanks for that update.

**Comment by grahamwhaley** [comment]
I should add - the tests now ran for me with `runc` btw - thanks for fixing that.

File: metrics/lib/common.bash [hunk 2]
```diff
@@ -83,7 +83,7 @@ function init_env()
 	runtime=$(docker info | grep -w "Default Runtime:" \
```
Comment (bwarden) [inline metrics/lib/common.bash]: Fixed

File: metrics/lib/common.bash [hunk 3]
```diff
@@ -163,10 +163,16 @@ function get_qemu_path(){
 
 # Checking that default runtime is appropriate
```
Comment (bwarden) [inline metrics/lib/common.bash]: Nitpicks welcome; fixed.

File: metrics/lib/common.bash (hunk 1 above)
Comment (bwarden) [inline metrics/lib/common.bash]: Ah, I need to make $found local... new push coming.

**Comment by grahamwhaley** [comment]
OK, this lgtm, but we failed (not our fault) both:
Travis: revision range issue
Semaphore: test failures, that are not related to this PR, so I think must be a transient thing that @chavafg can inform us about.

Let's work out the Semaphore thing, and then we may need a rebase etc. for the Travis. If all that fails, we could force merge, but hopefully we won't have to.

**Comment by bwarden** [comment]
I got Travis to pass, but Semaphore failed again.

**Comment by grahamwhaley** [comment]
@bwarden ta. I'm not surprised. I suspect this is something QA/CI broken that might need a fix before we can pass. @chavafg - I've not been tracking if we have any known pending major issues with the CI since we did the timeout change - is the below something you recognise, and is there an obvious fix?
```
...
[Fail] run container [It] with '[sh -c exit 15]' as workload 
/home/runner/workspace/src/github.com/clearcontainers/tests/functional/run_test.go:59

[Fail] run container [It] with '[sh -c exit 123]' as workload 
/home/runner/workspace/src/github.com/clearcontainers/tests/functional/run_test.go:59

Ran 27 of 27 Specs in 3.741 seconds
FAIL! -- 17 Passed | 10 Failed | 0 Pending | 0 Skipped --- FAIL: TestFunctional (3.74s)
FAIL

Ginkgo ran 1 suite in 5.843165275s
Test Suite Failed
make: *** [functional] Error 1
```

**Comment by sboeuf** [comment]
Those failures are weird because they are failing everywhere. Have you rebased your commit on top of origin/master after you fetched origin ? This could be the problem if you are missing some commits.

**Comment by GabyCT** [comment]
I do not think these failures are related with timeouts, if you check the log
ln -sf . vendor/src
GOPATH=/home/runner/workspace/src/github.com/clearcontainers/tests/vendor go build ./vendor/github.com/onsi/ginkgo/ginkgo
unlink vendor/src
./ginkgo functional/ -- -runtime cc-runtime -timeout 5
Running Suite: Functional Suite
===============================
Random Seed: 1502466224
Will run 27 of 27 specs

•••••••••••••••Running command '/usr/local/bin/cc-runtime [cc-runtime --debug --log /tmp/bundle473143467/log run --bundle /tmp/bundle473143467 --console  vDVV99DwA3TEqUSY0I2a]'
command failed error 'exit status 1'
/etc/clear-containers/configuration.toml: File does not exist: /usr/share/clear-containers/vmlinuz.container
Running command '/usr/local/bin/cc-runtime [cc-runtime list --quiet]'
command failed error 'exit status 1'

Do you have the latests commits?

**Comment by bwarden** [comment]
It has been a few days since I rebased... pushing now.

**Comment by grahamwhaley** [comment]
yay, CI worked - quick, let's merge it before it changes its mind....

**Pull request merged**